pkg/controller/argocdmetrics/argocd_metrics_controller.go

```diff
@@ -0,0 +1,409 @@
+	readRoleNameFormat        = "%s-read"
+	readRoleBindingNameFormat = "%s-prometheus-k8s-read-binding"
+	alertRuleName             = "gitops-operator-argocd-alerts"
+)
+
+var logs = logf.Log.WithName("controller_argocd_metrics")
+
+type ArgoCDMetricsReconciler struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// blank assignment to verify that ReconcileArgoCDRoute implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ArgoCDMetricsReconciler{}
+
+// Add creates a new ArgoCD Metrics Controller and adds it to the Manager.
+// The Manager will set fields on the Controller and start it when the
+// Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+func add(mgr manager.Manager, reconciler reconcile.Reconciler) error {
+	reqLogger := logs.WithValues()
+
+	reqLogger.Info("Creating ArgoCD metrics controller")
+	c, err := controller.New("argocd-metrics-controller", mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return err
+	}
+
+	reqLogger.Info("Watching for ArgoCD instance creation")
+	err = c.Watch(&source.Kind{Type: &argoapp.ArgoCD{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ArgoCDMetricsReconciler{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+func (r *ArgoCDMetricsReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := logs.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ArgoCD Metrics")
+
+	namespace := corev1.Namespace{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: request.Namespace}, &namespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Namespace not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "Error getting namespace",
+			"Namespace", request.Namespace)
+		return reconcile.Result{}, err
+	}
+
+	argocd := &argoapp.ArgoCD{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: request.Name, Namespace: request.Namespace}, argocd)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// ArgoCD not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "Error getting ArgoCD instsance")
+		return reconcile.Result{}, err
+	}
+
+	const clusterMonitoringLabel = "openshift.io/cluster-monitoring"
```

Comment:
This const could be promoted to package scope so that the test module uses it as well.

Reply:
I don't think it's a good practice to share constants between test and production code.  If the constant got accidentally changed in the production code, the test would still pass, but the code would fail when running for real.

Reply:
OK.  That's fair.